src/rustzcash.rs

```diff
@@ -301,6 +301,88 @@ pub extern "system" fn librustzcash_ivk_to_pkd(
     }
 }
 
+/// Test generation of commitment randomness
+#[test]
+fn test_gen_r() {
+    let r1 = [0u8; 32];
+    let r2 = [0u8; 32];
+
+    // Verify different r values are generated
+    assert_ne!(r1, r2);
+
+    // Verify r values are valid in the field
+    let mut repr = FsRepr::default();
+    repr.read_le(&r1[..]).expect("length is not 32 bytes");
+    let _ = Fs::from_repr(repr).unwrap();
+    repr.read_le(&r2[..]).expect("length is not 32 bytes");
+    let _ = Fs::from_repr(repr).unwrap();
+}
+
+/// Return 32 byte randomness, uniform, to be used for a Sapling commitment.
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_generate_r(result: *mut [c_uchar; 32]) {
+    // create random 64 byte buffer
+    let mut rng = OsRng::new().expect("should be able to construct RNG");
+    let mut buffer = [0u8; 64];
+    for i in 0..buffer.len() {
+        buffer[i] = rng.gen();
+    }
+
+    // reduce to uniform value
+    let r = <Bls12 as JubjubEngine>::Fs::to_uniform(&buffer[..]);
+    let result = unsafe { &mut *result };
+    r.into_repr()
+        .write_le(&mut result[..])
+        .expect("result must be 32 bytes");
+}
+
+/// Compute Sapling note commitment.
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_compute_cm(
+    diversifier: *const [c_uchar; 11],
+    pk_d: *const [c_uchar; 32],
+    value: uint64_t,
+    r: *const [c_uchar; 32],
+    result: *mut [c_uchar; 32],
+) -> bool {
+    let diversifier = sapling_crypto::primitives::Diversifier(unsafe { *diversifier });
+    let g_d = match diversifier.g_d::<Bls12>(&JUBJUB) {
+        Some(g_d) => g_d,
+        None => return false,
+    };
+
+    let pk_d = match edwards::Point::<Bls12, Unknown>::read(&(unsafe { &*pk_d })[..], &JUBJUB) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    let pk_d = match pk_d.as_prime_order(&JUBJUB) {
```

Comment:
The spec doesn't require pk<sub>d</sub> to be prime order in keys obtained from another source (pk<sub>d</sub> in a note is declared as having type KA<sup>Sapling</sup>.Public which is J, not J<sub>(r)</sub>). I could make it require this outside the circuit, but note that the circuit doesn't check it.

Reply:
@ebfull Thoughts?

Reply:
I've looked closely at the changes that would be needed to the spec, and I'm convinced that the spec should be changed to require pk<sub>d</sub> (and g<sub>d</sub>, but that's already implied) to be of prime order.